### PR TITLE
ci: enables clippy and rustfmt in compile-time-options build

### PR DESCRIPTION
**Pull request by mathetake**
Commit Message: ci: enables clippy and rustfmt in compile-time-options build
Additional Description:
With the dynamic module features being developed, enabling the standard
formatter and linter for Rust source code should be setup.
This commit enables clippy and rustfmt checks that run during the build time.

Risk Level: None
Testing:  N/A
Docs Changes: N/A
Release Notes:  N/A
Platform Specific Features:  N/A
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by mathetake** [comment]
@phlax could you take a look?

File: .bazelrc [hunk 1]
```diff
@@ -351,6 +351,12 @@ build:remote-clang-cl --config=remote-windows
 build:remote-clang-cl --config=clang-cl
 build:remote-clang-cl --config=rbe-toolchain-clang-cl
 
+# Enable auto rustfmt and clippy during build time of Rust source code.
+build:rust-format --aspects=@rules_rust//rust:defs.bzl%rust_clippy_aspect
+build:rust-format --output_groups=+clippy_checks
```
Comment (mathetake) [inline .bazelrc:356]: this is the officially recommended way to enable this https://bazelbuild.github.io/rules_rust/rust_clippy.html#setup

File: .bazelrc [hunk 2]
```diff
@@ -351,6 +351,12 @@ build:remote-clang-cl --config=remote-windows
...
+build:rust-format --aspects=@rules_rust//rust:defs.bzl%rustfmt_aspect
+build:rust-format --output_groups=+rustfmt_checks
```
Comment (mathetake) [inline .bazelrc:358]: also https://bazelbuild.github.io/rules_rust/rust_fmt.html#setup

**Review by mathetake — COMMENTED** [review]
added notes

**Review by phlax — COMMENTED** [review]
hi @mathetake 

iiuc clippy is a static analyzer so im wondering why this is being added to build tests

relatedly this would only run against test targets and not any actual code (not sure if that matters as only rust code currently is in tests)

i would have thought this would be better to add in the format prechecks - not sure compile-time-options is a good place, not least because it would run repeatedly against the same code

ultimately i think we do want this, but in a way that contributors can run quickly (ie only run this, no unnecessary toolchains/deps/etc), and for ci we ideally want to make use of the `--fix` with an uploaded diff to quickly resolve any failure

**Comment by mathetake** [comment]
> for ci we ideally want to make use of the --fix with an uploaded diff to quickly resolve any failure

Sounds good! Thank you for the suggestion. I will try to look into it

**Comment by phlax** [comment]
if you ping me on slack im happy to help figuring out where to put it

**Comment by mathetake** [comment]
so now i have two separate issues:
* for rustfmt, it actually can be run with `bazel run @rules_rust//:rustfmt` for all rust targets. However, it doesn't expose `--check` flag, so it cannot fail, but it simply formats all. If `--check` is exposed, it is actually possible to integrate with the current `//tools/code_format:check_format -- fix/check` stuff.
* for clippy, this is worse because it does't expose any "run" target, so the only way to enable this is add `build --aspects=@rules_rust//rust:defs.bzl%rustfmt_aspect` for builds. However, this is a linter, so it cannot work well with `fix` flag of  `//tools/code_format:check_format` because sometimes the linter error cannot be automatically fixed by nature.

what do you think @phlax

**Comment by phlax** [comment]
re the first if there is no failure, we can make any diff a fail, and upload etc

re second not sure - my main thought is that its not much code right now so probs i would go the path of just getting the clippy bin/toolchain and running that on files

imhe the story with bazel and linting is not great - for sure aspects are the ~right way to do it but i found for other formatters - ie python/shelleck/go and some others, that the far faster solution was to run them in a python proc as efficiently as poss (ie https://github.com/envoyproxy/toolshed/tree/main/envoy.code.check) 

i would be happy just getting that to run it tbh - if we can feed it the necessary binary it should be pretty easy

otherwise we can go the aspect way, but then we need to make sure we are doing the minimal amount of work and targetting just those aspects - so end users can run it ~quickly

**Comment by mathetake** [comment]
ok thanks for the comment. I think I would rather go with the explicit build target configurations for both rustfmt and clippy instead of trying to integrate them into check_format in a hacky way. As you suggested, there's not many code and won't be for a foreseeable future, so I would explicitly add the followings in each BUILD 

```
rustfmt_test(
    name = "hello_library_fmt",
    target = ":hello_lib",
)
rust_clippy(
    name = "hello_library_clippy",
    deps = [":hello_lib"],
)
```

until there are many rust codes and adding this across many places tedious. Thank you for you comments @phlax anyways, and i am closing this for now!

**Pull request closed without merging**